x-pack/legacy/plugins/lens/server/routes/index_stats.ts

```diff
@@ -164,3 +275,170 @@ export function recursiveFlatten(
   });
   return returnTypes;
 }
+
+export async function getNumberHistogram(
```

Comment:
The interleaving of code and type definitions interfered with my reading of these functions. I'd suggest breaking the type definitions out. Also, you can make the search function parameterized, which helps a bit:

```js
type MinMaxResult = AggregationSearchResponse<
  unknown,
  {
    body: {
      aggs: {
        sample: {
          sampler: {};
          aggs: {
            min_value: {
              min: { field: string };
            };
            max_value: {
              max: { field: string };
            };
          };
        };
      };
    };
  }
>;

type NumericHistogramResult = AggregationSearchResponse<
  unknown,
  {
    body: {
      aggs: {
        sample: {
          sampler: {};
          aggs: {
            histo: {
              histogram: { field: string; interval: number };
            };
          };
        };
      };
    };
  }
>;

export async function getNumberHistogram(
  aggSearchWithBody: <T>(body: unknown) => Promise<T>,
  field: { name: string; type: string; esTypes?: string[] }
) {
  const minMaxResult = await aggSearchWithBody<MinMaxResult>({
    sample: {
      sampler: { shard_size: 5000 },
      aggs: {
        min_value: {
          min: { field: field.name },
        },
        max_value: {
          max: { field: field.name },
        },
        top_values: {
          terms: { field: field.name, size: 10 },
        },
      },
    },
  });

  const minValue = minMaxResult.aggregations!.sample.min_value.value;
  const maxValue = minMaxResult.aggregations!.sample.max_value.value;
  const terms = minMaxResult.aggregations!.sample.top_values;

  let histogramInterval = (maxValue! - minValue!) / 10;

  if (Number.isInteger(minValue!) && Number.isInteger(maxValue!)) {
    histogramInterval = Math.ceil(histogramInterval);
  }

  if (histogramInterval === 0) {
    return {
      topValues: terms,
      histogram: { buckets: [], docCount: minMaxResult.aggregations!.sample.doc_count },
    };
  }

  const histogramResult = await aggSearchWithBody<NumericHistogramResult>({
    sample: {
      sampler: { shard_size: 5000 },
      aggs: {
        histo: {
          histogram: {
            field: field.name,
            interval: histogramInterval,
          },
        },
      },
    },
  });

  return {
    // Probably need to convert the histogram's props from snake_case to camelCase here...
    histogram: histogramResult.aggregations!.sample,
    topValues: terms,
  };
}
```

Reply:
I've actually done even better than this, check it out!

Reply:
I have one suggestion to reduce the number of casts and unknowns by unifying it in one place: Instead of passing in a function to do the search, the individual getter functions take a search dependency object that they pass to a module scope search function that does the cast once:
```ts
// post handler

  // collect everything necessary to do searches 
  const searchDeps = { requestClient, index: req.params.indexPatternTitle, filters };

  if (field.type === 'number') {
    return res.ok({
      // pass it to the actual getter
      body: await getNumberHistogram(searchDeps, field),
  });

//...

// these are the dependencies necessary for searches
interface SearchDependencies {
  requestClient: ScopedClusterClient;
  index: string;
  filters: unknown;
}

// search function itself has a generic type that makes it possible to infer
// the return value from the aggs parameter
function search<T>({ requestClient, index, filters }: SearchDependencies, aggs: T) {
  return requestClient.callAsCurrentUser('search', {
    index,
    body: {
      query: filters,
      aggs,
    },
    // The hits total changed in 7.0 from number to object, unless this flag is set
    // this is a workaround for elasticsearch response types that are from 6.x
    restTotalHitsAsInt: true,
    size: 0,
   // still have to cast here because callAsCurrentUser won't do it for us
  }) as Promise<AggregationSearchResponse<unknown, { body: { aggs: T } }>>;
}

// ...

export async function getNumberHistogram(
  searchDeps: SearchDependencies,
  field: { name: string; type: string; esTypes?: string[] }
): Promise<FieldStatsResponse> {
  // ...
  // can use search here without any casts, everything is inferred from searchBody
  const minMaxResult = await search(searchDeps, searchBody);
  // ...
}
```